Switched BOM pom.xml to v.r.* ranges for included specs to allow for security fixes etc.

hutchig opened the pull request:
Signed-off-by: Gordon Hutchison <Gordon.Hutchison@gmail.com>

As requested in the Oct 8th Architecture call [minutes](https://docs.google.com/document/d/1uAKRymdAL5Wj1KADzrYKZzSnAFsq74lBfj8lxKQnI6I/edit)
This PR switches the BOM to allow ranges for the third digit (second decimal) version numbers
for included specs.

So the POM still specifies particular X.Y ranges but will pick up the latest Z in X.Y.Z
(a change to X may be a breaking change, a change to Y alone
should be backwards compatible with existing users of any API).
So for X.Y.Z a variation in Z alone should have identical API but allow for
security and maintenance fixes from individual specs to be picked up.

Both Jakarta and Microprofile use this convention.

See [here](https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#CJHDEHAB), for example for the Maven syntax used.

kenfinnigan commented on pom.xml:
```diff
@@ -99,21 +99,21 @@
-        <jsonb-version>1.0</jsonb-version>
-        <jsonp-version>1.1</jsonp-version>
-        <annotation-version>1.3</annotation-version>
+        <cdi-version>[2.0,2.1)</cdi-version>
```
I don't think we discussed version ranges for base specifications coming from Java EE/Jakarta EE. I'd be wary of adding it for the base specs

hutchig replied:
Yes, I wondered about this too. I had a rummage then wrote:
"Both Jakarta and Microprofile use this convention."
(because they do ) but I agree it is beyond what we discussed - if the concensus is that, for example,  2.0.*
benefits are not worth the risk for Jakarta then I can remove it but 
I think the other way. I don't think Jakarta spec leads are going to make breaking
spec changes as 3rd digit changes and I don't think we should treat Jakarta as less
trust worthy in any governance issue like that. I don't want to anyway.

hutchig replied:
I have removed version changes from the base javax specs.

hutchig commented:
See https://github.com/eclipse/microprofile/pull/134#discussion_r335565962 for some interesting points from @joakime

hutchig commented:
HOLD added so we can digest https://github.com/eclipse/microprofile/pull/132#issuecomment-542783729

hutchig commented:
Please could you review this @Emily-Jiang ?

hutchig commented:
I have removed the HOLD label as I have thoroughly investigated the exact
maven range lower bounds and added the `.alpha` decorator to the non inclusive
upper bounds of the ranges as  by definition that is the very very lowest version
of the 'next' version up.

Azquelt commented:
I think the ranges as they are now would still pick up a micro version release candidate (e.g. 1.0.1-RC1). Not sure if that's an issue or not.

hutchig commented:
I believe the way round that would be to use a rule to ignore such:

```xml

<plugin>
    <groupId>org.codehaus.mojo</groupId>
    <artifactId>versions-maven-plugin</artifactId>
    <version>2.1</version>
    <configuration>
        <rulesUri>file:///${session.executionRootDirectory}/rules.xml</rulesUri>
    </configuration>
</plugin>
.... where rules.xml is 
<ruleset 
  comparisonMethod="maven"
  xmlns="http://mojo.codehaus.org/versions-maven-plugin/rule/2.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://mojo.codehaus.org/versions-maven-plugin/rule/2.0.0 http://www.mojohaus.org/versions-maven-plugin/xsd/rule-2.0.0.xsd">

  <rule groupId="org.eclipse.microprofile">
    <ignoreVersions>
      <ignoreVersion type="regex">.*[-_\.](alpha|Alpha|ALPHA|b|beta|Beta|BETA|rc|‌​RC|M|EA)[-_\.]?[0-9]*</ignoreVersion>
    </ignoreVersions>
  </rule>
</ruleset>
```

hutchig commented:
I put the HOLD back on as I want to investigate how this rule works on transitive dependencies (i.e distributing the rules so they are active later. This is possible I believe: https://stackoverflow.com/questions/26120704/maven-versions-plugin-reference-a-rule-xml-from-a-maven-dependency

kwsutter commented:
I read the [minutes from the architecture call](https://docs.google.com/document/d/1uAKRymdAL5Wj1KADzrYKZzSnAFsq74lBfj8lxKQnI6I/edit), and I'm still confused as to what we are trying to accomplish.  This pom file is really only used when releasing a new version of the MicroProfile platform.  If a component has a security patch (or whatever), then they should just submit a PR against the pom.xml to get the upper bounds updated for their component.  I don't think we should be automatically picking up third digit updates without some indication from the component that they should be picked up for the platform.  An explicit PR would accomplish that.

hutchig commented:
I am fine with what Kevin said. It was an interesting PR for me to do but I don't think this is the
time or place to do a custom ruleset etc. So if no one objects by Tuesday's call I will close
this PR then.
